src/client/dfuse/pil4dfs/int_dfs.c

```diff
@@ -622,45 +623,45 @@ discover_daos_mount(void)
 	num_dfs++;
 }
 
+#define MNT_TYPE_FUSE	"fuse.daos"
 static void
 discover_dfuse(void)
 {
 	FILE          *fp;
-	const char     mnt_type_fuse[] = "fuse.daos";
 	struct mntent *fs_entry;
-	char          *pt_end;
+	struct dfs_mt *pt_dfs_mt;
 
 	num_dfs = 0;
 
 	fp = setmntent("/proc/self/mounts", "r");
-	/* daos_init() is not called yet, so we can not use D_WARN() etc. */
 
 	if (fp == NULL) {
-		D_ERROR("failed to open file: /proc/self/mounts\n");
+		D_ERROR("failed to open /proc/self/mounts: %d (%s)\n", errno, strerror(errno));
 		return;
 	}
 
 	while ((fs_entry = getmntent(fp)) != NULL) {
-		if (strncmp(fs_entry->mnt_type, mnt_type_fuse, sizeof(mnt_type_fuse)) == 0) {
-			dfs_list[num_dfs].dfs_dir_hash = NULL;
-			dfs_list[num_dfs].len_fs_root  = strnlen(fs_entry->mnt_dir, DFS_MAX_PATH);
-			if (dfs_list[num_dfs].len_fs_root >= DFS_MAX_PATH) {
-				D_WARN("mnt_dir[] is too long! Skip this entry.");
+		pt_dfs_mt = &dfs_list[num_dfs];
+		if (strncmp(fs_entry->mnt_type, MNT_TYPE_FUSE, sizeof(MNT_TYPE_FUSE)) == 0) {
+			pt_dfs_mt->dfs_dir_hash = NULL;
+			pt_dfs_mt->len_fs_root  = strnlen(fs_entry->mnt_dir, DFS_MAX_PATH);
+			if (pt_dfs_mt->len_fs_root >= DFS_MAX_PATH) {
+				D_DEBUG(DB_ANY, "mnt_dir[] is too long! Skip this entry.\n");
 				continue;
 			}
-			atomic_store_relaxed(&dfs_list[num_dfs].inited, 0);
-			dfs_list[num_dfs].pool         = NULL;
-			pt_end = stpncpy(dfs_list[num_dfs].fs_root, fs_entry->mnt_dir,
-					 DFS_MAX_PATH - 1);
-			if ((long int)(pt_end - dfs_list[num_dfs].fs_root) >= (DFS_MAX_PATH - 1)) {
-				D_WARN("fs_root[] is too long. Skip this entry.");
+			atomic_init(&pt_dfs_mt->inited, 0);
+			pt_dfs_mt->pool         = NULL;
+			/* The length of fs_entry->mnt_dir[] has been checked already. */
+			D_ALLOC(pt_dfs_mt->fs_root, DFS_MAX_PATH);
```

Comment:
At this point you know the length of the string so you can just use strndup() to allocate the space needed.
```suggestion
			D_STRNDUP(pt_dfs_mt->fs_root, fs_entry->mnt_dir, pt_dfs_mt->len_fs_root);
```

Reply:
I tried D_STRNDUP() and got NLT errors. Is this because D_FREE() is called in the destructor function?
I use strndup() and free() in the recent commit to avoid NLT failure.